kv/check.go

```diff
@@ -0,0 +1,704 @@
+		return err
+	}
+	return nil
+}
+
+// UnexpectedCheckError is used when the error comes from an internal system.
+func UnexpectedCheckError(err error) *influxdb.Error {
+	return &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  fmt.Sprintf("unexpected error retrieving check's bucket; Err %v", err),
+		Op:   "kv/checkBucket",
+	}
+}
+
+// UnexpectedCheckIndexError is used when the error comes from an internal system.
+func UnexpectedCheckIndexError(err error) *influxdb.Error {
+	return &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  fmt.Sprintf("unexpected error retrieving check's index bucket; Err %v", err),
+		Op:   "kv/checkIndex",
+	}
+}
+
+func (s *Service) checksBucket(tx Tx) (Bucket, error) {
+	b, err := tx.Bucket(checkBucket)
+	if err != nil {
+		return nil, UnexpectedCheckError(err)
+	}
+
+	return b, nil
+}
+
+func (s *Service) checksIndexBucket(tx Tx) (Bucket, error) {
+	b, err := tx.Bucket(checkIndex)
+	if err != nil {
+		return nil, UnexpectedCheckIndexError(err)
+	}
+
+	return b, nil
+}
+
+// FindCheckByID retrieves a check by id.
+func (s *Service) FindCheckByID(ctx context.Context, id influxdb.ID) (influxdb.Check, error) {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	var c influxdb.Check
+	var err error
+
+	err = s.kv.View(ctx, func(tx Tx) error {
+		chk, pe := s.findCheckByID(ctx, tx, id)
+		if pe != nil {
+			err = pe
+			return err
+		}
+		c = chk
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (s *Service) findCheckByID(ctx context.Context, tx Tx, id influxdb.ID) (influxdb.Check, error) {
+	span, _ := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	encodedID, err := id.Encode()
+	if err != nil {
+		return nil, &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Err:  err,
+		}
+	}
+
+	bkt, err := s.checksBucket(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := bkt.Get(encodedID)
+	if IsNotFound(err) {
+		return nil, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  "check not found",
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := check.UnmarshalJSON(v)
+	if err != nil {
+		return nil, &influxdb.Error{
+			Err: err,
+		}
+	}
+
+	return c, nil
+}
+
+// FindCheck retrives a check using an arbitrary check filter.
+// Filters using ID, or OrganizationID and check Name should be efficient.
+// Other filters will do a linear scan across checks until it finds a match.
+func (s *Service) FindCheck(ctx context.Context, filter influxdb.CheckFilter) (influxdb.Check, error) {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	var c influxdb.Check
+	var err error
+
+	if filter.ID != nil {
+		c, err = s.FindCheckByID(ctx, *filter.ID)
+		if err != nil {
+			return nil, &influxdb.Error{
+				Err: err,
+			}
+		}
+		return c, nil
+	}
+
+	err = s.kv.View(ctx, func(tx Tx) error {
+		if filter.Org != nil {
+			o, err := s.findOrganizationByName(ctx, tx, *filter.Org)
```

Comment:
line 131 suggests that orgID lookup will be efficient but it looks like this function searches for org Name only?

Reply:
Good point, I think it's related the org and orgID filter issue, we will make update for all them in one pr.